File: src/main/java/com/tw/pos/PosMachine.java
package com.tw.pos;

import com.tw.pos.entity.CartItem;
import com.tw.pos.entity.Goods;
import com.tw.pos.entity.GoodsRepository;

import java.util.List;


public class PosMachine {

    /*public double calculate(CartItem cartItem) {
        double sum = 0;
        for (int i = 1; i <= cartItem.getNumber(); i++) {
            if (i % 2 == 0) {
                sum += getGoodsPrice(cartItem.getBarcode());
            } else {
                Goods goods = getGoods(cartItem.getBarcode());
                if (goods.getDiscountPromotions().size() == 0){
                    sum += goods.getPrice();
                } else {
                    for (DiscountPromotion discountPromotion : goods.getDiscountPromotions()){
                        if (discountPromotion instanceof SecondHalf) {
                            sum += goods.getPrice();
                        } else {
                            discountPromotion.apply();
                            sum += goods.getPrice();
                            discountPromotion.cancle();
                            discountPromotion.prepare();
                        }
                    }
                }
            }
        }
        return sum;
        //return cartItem.getNumber() * getGoodsPrice(cartItem.getBarcode());
    }*/
    public double calculate(CartItem cartItem){
        return cartItem.getNumber() * getGoodsPrice(cartItem.getBarcode());
    }

    public double calculate(List<CartItem> cartItems) {
        double sum = 0;
        for (CartItem cartItem : cartItems){
            sum += calculate(cartItem);
        }
        return sum;
    }

    public Goods getGoods(String barcode){
        try {
            return GoodsRepository.getGoodsByBarcode(barcode);
        }catch (Exception e){
            throw new IllegalArgumentException("invalid goods");
        }
    }

    public double getGoodsPrice(String barcode){
        Goods goods = getGoods(barcode);
        goods.useDiscountPromotion();
        return goods.getPrice();

    }
}

File: js/posMachine.js
var PosMachine = function (allGoods, discountManager) {
    this.allGoods = allGoods;
    this.discountManager = discountManager;
};

PosMachine.prototype.calculate = function (cartItem) {
    var goods = this.getGoods(cartItem.barcode);

    var discounts = this.discountManager.discountMap[goods.barcode];
    if (discounts == null || discounts.size == 0){
        return goods.price * cartItem.number;
    }

    var originPrice = goods.price;
    var totalPrice = 0;
    for (var i = 0, discount; discount = discounts[i++];){
        totalPrice = discount.apply(cartItem, originPrice);
        originPrice = discount.getDiscountPrice(originPrice);
    }
    return totalPrice;
};

PosMachine.prototype.calculateCartItems = function (cartItems) {
    var sum = 0;
    for (var i = 0, cartItem; cartItem = cartItems[i++];) {
        sum += this.calculate(cartItem);
    }
    return sum;
};

PosMachine.prototype.getGoods = function (barcode) {
    for (var i = 0, goods; goods = this.allGoods[i++];) {
        if (goods.barcode == barcode) {
            return goods;
        }
    }
    throw new Error("invalid goods");
};